### ISSUE TITLE
userdbd kills itself after git+ssh logins frequently

### ISSUE THREAD

**Issue by SjonHortensius** [systemd/systemd#27028]
### systemd version the issue has been seen with

253

### Used distribution

archlinux

### Linux kernel version used

6.2.2-arch1-1

### CPU architectures issue was seen on

x86_64

### Expected behaviour you didn't see

* userdbd should have no issues with high frequency normal logins
* userdbd should not fail restarting itself

### Unexpected behaviour you saw

on a machine which is used as a git server (through ssh) sessions are rapidly spawned. This apparently stresses userdbd so much it fails (these 7 logins happen within one second):

```
systemd[1]: Started Session 179906 of User git.
systemd-logind[3799586]: New session 179907 of user git.
systemd[1]: Started Session 179907 of User git.
systemd-logind[3799586]: Removed session 179900.
systemd[1]: session-179904.scope: Deactivated successfully.
systemd-logind[3799586]: New session 179908 of user git.
systemd[1]: Started Session 179908 of User git.
systemd-logind[3799586]: New session 179910 of user git.
systemd[1]: Started Session 179910 of User git.
systemd-logind[3799586]: New session 179909 of user git.
systemd[1]: Started Session 179909 of User git.
systemd-logind[3799586]: New session 179911 of user git.
systemd[1]: Started Session 179911 of User git.
systemd-logind[3799586]: Removed session 179904.
systemd-logind[3799586]: New session 179913 of user git.
systemd[1]: Started Session 179913 of User git.
systemd-logind[3799586]: New session 179912 of user git.
systemd[1]: Started Session 179912 of User git.
systemd[1]: session-179908.scope: Deactivated successfully.
systemd-logind[3799586]: Removed session 179908.
systemd[1]: session-179906.scope: Deactivated successfully.
systemd[1]: session-179913.scope: Deactivated successfully.
systemd[1]: session-179911.scope: Deactivated successfully.
systemd[1]: session-179912.scope: Deactivated successfully.
systemd[1]: session-179909.scope: Deactivated successfully.
systemd[1]: session-179907.scope: Deactivated successfully.
systemd[1]: session-179910.scope: Deactivated successfully.
systemd-logind[3799586]: Removed session 179906.
systemd-logind[3799586]: Removed session 179913.
systemd-logind[3799586]: Removed session 179911.
systemd-logind[3799586]: Removed session 179912.
systemd-logind[3799586]: Removed session 179909.
systemd-logind[3799586]: Removed session 179907.
systemd-logind[3799586]: Removed session 179910.
systemd[1]: Looping too fast. Throttling execution a little.
systemd-userdbd[1545485]: Worker threads requested too frequently, something is wrong.
systemd[1]: systemd-userdbd.service: Deactivated successfully.
systemd[1]: systemd-userdbd.service: Consumed 4min 44.112s CPU time, no IO, no IP traffic.
systemd[1]: Starting User Database Manager...
systemd[1]: Started User Database Manager.
systemd-userdbd[2006053]: Worker threads requested too frequently, something is wrong.
systemd[1]: systemd-userdbd.service: Deactivated successfully.
systemd[1]: Starting User Database Manager...
systemd[1]: Started User Database Manager.
systemd-userdbd[2006108]: Worker threads requested too frequently, something is wrong.
systemd[1]: systemd-userdbd.service: Deactivated successfully.
systemd[1]: Starting User Database Manager...
systemd[1]: Started User Database Manager.
systemd-userdbd[2006161]: Worker threads requested too frequently, something is wrong.
systemd[1]: systemd-userdbd.service: Deactivated successfully.
systemd[1]: Starting User Database Manager...
systemd[1]: Started User Database Manager.
systemd-userdbd[2006214]: Worker threads requested too frequently, something is wrong.
systemd[1]: systemd-userdbd.service: Deactivated successfully.
systemd[1]: Starting User Database Manager...
systemd[1]: Started User Database Manager.
systemd-userdbd[2006266]: Worker threads requested too frequently, something is wrong.
systemd[1]: systemd-userdbd.service: Deactivated successfully.
systemd[1]: systemd-userdbd.service: Start request repeated too quickly.
systemd[1]: systemd-userdbd.service: Failed with result 'start-limit-hit'.
systemd[1]: Failed to start User Database Manager.
```

### Steps to reproduce the problem

this is normal git usage - with `KillUserProcesses=yes` in logind.conf or `loginctl` would show 8192 "active" sessions and refuse to work normally, which imo is another bug

**Comment by vcaputo** [comment]
hmm yeah just a casual glance at the code it seems like a fairly obvious bug:

This block will request workers as fast as requests come in:
https://github.com/systemd/systemd/blob/v253/src/userdb/userwork.c#L543-L561

But the recipient process of the SIGUSR2 spawning the workers just exits when the ratelimit triggers, which apparently wasn't an expected normal behavior when authored:
https://github.com/systemd/systemd/blob/v253/src/userdb/userdbd-manager.c#L234-L239

**Comment by poettering** [comment]
Fix waiting in → #28129
